src/borg/archiver.py

```diff
@@ -652,126 +709,82 @@ def _process(self, *, path, parent_fd=None, name=None,
                         if tag_names:
                             # if we are already recursing in an excluded dir, we do not need to do anything else than
                             # returning (we do not need to archive or recurse into tagged directories), see #3991:
-                            if not recurse_excluded_dir:
-                                if keep_exclude_tags and not dry_run:
-                                    fso.process_dir(path=path, fd=child_fd, st=st)
-                                    for tag_name in tag_names:
-                                        tag_path = os.path.join(path, tag_name)
-                                        self._process(path=tag_path, parent_fd=child_fd, name=tag_name,
-                                                      fso=fso, cache=cache, matcher=matcher,
-                                                      exclude_caches=exclude_caches, exclude_if_present=exclude_if_present,
-                                                      keep_exclude_tags=keep_exclude_tags, skip_inodes=skip_inodes,
-                                                      restrict_dev=restrict_dev, read_special=read_special, dry_run=dry_run)
-                                self.print_file_status('x', path)
+                            if recurse_excluded_dir:
+                                return
+                            if keep_exclude_tags:
+                                status = self._process_any(path=path, parent_fd=parent_fd, name=name, st=st, fso=fso,
+                                                      cache=cache, read_special=read_special, dry_run=dry_run)
+                                for tag_name in tag_names:
+                                    tag_path = os.path.join(path, tag_name)
+                                    self._rec_walk(
+                                            path=tag_path, parent_fd=child_fd, name=tag_name, fso=fso, cache=cache,
+                                            matcher=matcher, exclude_caches=exclude_caches, exclude_if_present=exclude_if_present,
+                                            keep_exclude_tags=keep_exclude_tags, skip_inodes=skip_inodes,
+                                            restrict_dev=restrict_dev, read_special=read_special, dry_run=dry_run)
+                            self.print_file_status('x', path)
                             return
-                    if not dry_run:
-                        if not recurse_excluded_dir:
-                            status = fso.process_dir(path=path, fd=child_fd, st=st)
-                    if recurse:
+
+                    if not recurse_excluded_dir:
+                        status = self._process_any(path=path, parent_fd=parent_fd, name=name, st=st, fso=fso,
+                                                   cache=cache, read_special=read_special, dry_run=dry_run)
+
+                    if right_dev:
```

Comment:
at this place, i find recurse an even better name than right_dev.

if it is true, we scan the directory (== recurse). wanting or not wanting to recurse could also have other reasons than it being the right device or not.

Reply:
For my defense: `recurse` isn't really explicit about what makes that variable be true/false, whereas the fact that it's used to decide on recursion should be obvious from the surrounding of it's use. Also i started using the value of `right_dev` with an interpretation unrelated to recursing or not (like line 764, arguable line )... Anyway you're gonna read this code more than me, i'll do whatever you say!

Reply:
thanks